File: seismic_benchmarks/laplacian/2D/laplacian-2d-mpi.f90
!*%****************************************************************************80
!  Code: 
!   laplacian-2d-mpi.f90
!
!  Purpose:
!   Implements sample 2D Laplacian Method in fortran using MPI.
!
!  Modified:
!   Aug 17 2020 10:57 
!
!  Author:
!   Murilo Boratto <murilo.boratto 'at' fieb.org.br>
!   Reynam da Cruz Pestana <reynam 'at' ufba.br>
!
!  How to Compile:
!   mpifort laplacian-2d-mpi.f90 -o object
!
!  How to Execute: 
!   mpirun -np 4 ./object                             
!     
!  Comments: 
!   1) Simple Testbed with size problem n = m = 8.                              
!****************************************************************************80*/ 

PROGRAM laplacian_2d_mpi

	    USE MPI

		IMPLICIT NONE

		INTEGER(kind=4) :: myid   ! number associated with each processor, starting from zero
		INTEGER(kind=4) :: nprocs ! number of processors that will be used
		INTEGER(kind=4) :: ierr   ! error
		INTEGER :: n, m = 8
		REAL :: dx, dz
		REAL*8, DIMENSION(:,:), ALLOCATABLE :: a, b, c, sx, sz
		INTEGER :: status(MPI_STATUS_SIZE)
		INTEGER :: jsta,jend,jsta2,jend2,inext,iprev, isend1, isend2, irecv1, irecv2
		INTEGER :: i, j

		! Initializing MPI
		CALL MPI_init(ierr)
		CALL MPI_Comm_size(MPI_COMM_WORLD, nprocs, ierr)
		CALL MPI_Comm_rank(MPI_COMM_WORLD, myid, ierr)

		IF (myid.eq.0) THEN
		   n = m
		END IF

		CALL MPI_BCAST(n,1,MPI_INTEGER,0,MPI_COMM_WORLD,ierr)
		CALL MPI_BCAST(m,1,MPI_INTEGER,0,MPI_COMM_WORLD,ierr)

		ALLOCATE(a(m,n),b(m,n),c(m,n),sx(m,n),sz(m,n))

		a  = 0.
		b  = 0.
		c  = 0.
		sx = 0.
		sz = 0.
		dx = 1.
		dz = 1.

		CALL PARA_RANGE_1(1, n, nprocs, myid, jsta, jend)

		jsta2 = jsta
		jend2 = jend

		IF(myid.eq.0) THEN
		   jsta2 = 2
		END IF

		IF(myid.eq.(nprocs-1)) THEN
		   jend2 = n-1
		END IF

		inext = myid + 1
		iprev = myid - 1

		IF(myid.eq.(nprocs-1)) THEN
		   inext = MPI_PROC_NULL
		END IF

		IF(myid.eq.0) THEN
		   iprev = MPI_PROC_NULL
		END IF

		DO i = 1, m
		  DO j = jsta, jend
			a(i,j) = (i + j + 2) * 1.
		  END DO
		END DO

                CALL PRINT_MATRIX(a, n)

		CALL MPI_ISEND(a(1,jend),   m, MPI_REAL8, inext, 1, MPI_COMM_WORLD, isend1, ierr)
		CALL MPI_ISEND(a(1,jsta),   m, MPI_REAL8, iprev, 1, MPI_COMM_WORLD, isend2, ierr)
		CALL MPI_IRECV(a(1,jsta-1), m, MPI_REAL8, iprev, 1, MPI_COMM_WORLD, irecv1, ierr)
		CALL MPI_IRECV(a(1,jend+1), m, MPI_REAL8, inext, 1, MPI_COMM_WORLD, irecv2, ierr)

		CALL MPI_WAIT(isend1, status, ierr)
		CALL MPI_WAIT(isend2, status, ierr)
		CALL MPI_WAIT(irecv1, status, ierr)
		CALL MPI_WAIT(irecv2, status, ierr)

		DO j = jsta2,jend2
		   DO i = 2, m - 1
		     sx(i,j) = a(i-1,j) + a(i+1,j) + 2*a(i,j)
		     sz(i,j) = a(i,j-1) + a(i,j+1) + 2*a(i,j)
	              c(i,j) = sx(i,j)/(dx**2) + sz(i,j)/(dz**2)
		   END DO
		END DO

		CALL MPI_REDUCE(c, b, m*n, MPI_REAL8, MPI_SUM, 0, MPI_COMM_WORLD, ierr)

		IF (myid.eq.0) THEN
	           CALL PRINT_MATRIX(b,n)
		!  OPEN(unit=10, file='laplaciano_fortran.dat', ACCESS='direct', recl=8*n*m, status='UNKNOWN')
		!  WRITE(10,rec=1)real(b,kind=8)
		END IF

		DEALLOCATE (a, b, c, sx, sz)

                ! Finalizing MPI
	        CALL MPI_FINALIZE(ierr)

	        STOP

END PROGRAM laplacian_2d_mpi

! --------------------------------------------------------------------- !


SUBROUTINE PARA_RANGE_1(n1,n2,nprocs,myid,jsta,jend)

	INTEGER :: iwork1, iwork2, n1, n2, nprocs, myid, jsta, jend

	iwork1 = (n2-n1+1)/nprocs
	iwork2 = mod(n2-n1+1,nprocs)
	jsta   = myid*iwork1+n1+min(myid,iwork2)
	jend   = jsta+iwork1-1

	IF (iwork2.gt.myid) THEN
	  jend = jend + 1
	END IF

END SUBROUTINE PARA_RANGE_1


SUBROUTINE PRINT_MATRIX(a,n)

	INTEGER::n
	REAL*8::a(n,n) 

	DO i=1,n
          PRINT '(20f6.2)', a(i,1:n)
        ENDDO

        WRITE(*,*) ' '

END SUBROUTINE PRINT_MATRIX

File: seismic_benchmarks/laplacian/2D/laplacian-2d-mpi.c
/*%****************************************************************************80
!  Code: 
!   laplacian-2d-mpi.c
!
!  Purpose:
!   Implements sample 2D Laplacian Method in C/C++ code using MPI.
!
!  Modified:
!   Aug 18 2020 10:57 
!
!  Author:
!   Murilo Boratto <murilo.boratto 'at' fieb.org.br>
!   Reynam da Cruz Pestana <reynam 'at' ufba.br>
!
!  How to Compile:
!   mpicc aplacian-2d-mpi.c -o object -lm
!
!  How to Execute: 
!   mpirun -np 4 ./object
!
!  Comments: 
!   1) Simple Testbed with size problem n = m = 8. 
!****************************************************************************80*/

#include <mpi.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <math.h>

void show_matrix(double *a, int n){

   int i, j;

   for(i = 0; i < n; i++){
      for(j = 0; j < n; j++){
        printf("%1.2f\t", a[i + j*n]);
      }
    printf("\n");
   }

   printf("\n");

}/*show_matrix*/

void show_vector(int *a, int n){

   int i;

   for(i = 0; i < n; i++)
     printf("%d\t", a[i]);

   printf("\n\n");

}/*show_vector*/

void PARA_RANGE_1(int n1,int n2, int nprocs, int myid, int jsta, int jend, int *vector_return){

	int iwork1 = (n2 - n1 + 1) / nprocs;
	int iwork2 = (n2 - n1 + 1) % nprocs;

	jsta   = (myid * iwork1) + n1 + fmin(myid, iwork2);
	jend   = jsta + iwork1 - 1;

	if (iwork2 > myid)
	 jend = jend + 1;

        vector_return[0] = jsta;
        vector_return[1] = jend;

} /*PARA_RANGE_1*/


int main(int argc, char *argv[]){

   int n = 8;
   int m = n;
   int myid;
   int nprocs;
   double dx, dz;
   double *a, *b, *c, *sx, *sz;
   int jsta, jend, jsta2, jend2, inext, iprev ;
   int i, j;
   int *vector_return = (int *) calloc (2, sizeof(int));;
   FILE *file;

   MPI_Init (&argc, &argv);
   MPI_Comm_rank (MPI_COMM_WORLD, &myid);
   MPI_Comm_size (MPI_COMM_WORLD, &nprocs);
   MPI_Status status;
   MPI_Request isend1, isend2, irecv1, irecv2;

   if (myid == 0)
  	  m = n;  

   MPI_Bcast(&n, 1, MPI_INTEGER, 0, MPI_COMM_WORLD);
   MPI_Bcast(&m, 1, MPI_INTEGER, 0, MPI_COMM_WORLD);

   a  =  (double*) calloc (m * n, sizeof(double));
   b  =  (double*) calloc (m * n, sizeof(double));
   c  =  (double*) calloc (m * n, sizeof(double));
   sx =  (double*) calloc (m * n, sizeof(double));
   sz =  (double*) calloc (m * n, sizeof(double));

   dx = 1.;
   dz = 1.;

   PARA_RANGE_1(1, n, nprocs, myid, jsta, jend, vector_return);

   jsta = vector_return[0];
   jend = vector_return[1];

   jsta2 = jsta;
   jend2 = jend;

  if(myid == 0)
    jsta2 = 2;

   if(myid == (nprocs-1))
     jend2 = n - 1;

   inext = myid + 1;
   iprev = myid - 1;

   if(myid == ( nprocs - 1 ))
     inext = MPI_PROC_NULL;

   if(myid == 0)
     iprev = MPI_PROC_NULL;

   /*Population of the matrix*/
   for(i = 0; i < m; i++)
     for(j = jsta-1; j < jend; j++)
	 a[i + j* n] = (i + j + 2) * 1.;  /*stored dates in column (major column)*/
 
   MPI_Barrier(MPI_COMM_WORLD);

   if(myid == 0){ /*testbed*/
     show_matrix(a, n);
   }

   MPI_Barrier(MPI_COMM_WORLD); /*testbed*/

   if(myid == 1){ 
     show_matrix(a, n);
   }

   MPI_Barrier(MPI_COMM_WORLD); 

   if(myid == 2){ 
      show_matrix(a, n);
   }

   MPI_Barrier(MPI_COMM_WORLD); 

   if(myid == 3){ 
     show_matrix(a, n);
   }

   MPI_Isend(&a[0 + (jend - 1)   * n]  ,   m, MPI_REAL8, inext, 1, MPI_COMM_WORLD, &isend1);
   MPI_Isend(&a[0 + (jsta - 1)   * n]  ,   m, MPI_REAL8, iprev, 1, MPI_COMM_WORLD, &isend2);
   MPI_Irecv(&a[0 + (jsta - 2)   * n]  ,   m, MPI_REAL8, iprev, 1, MPI_COMM_WORLD, &irecv1);
   MPI_Irecv(&a[0 + (jend)       * n]  ,   m, MPI_REAL8, inext, 1, MPI_COMM_WORLD, &irecv2);

   MPI_Wait(&isend1, &status);
   MPI_Wait(&isend2, &status);
   MPI_Wait(&irecv1, &status);
   MPI_Wait(&irecv2, &status);

   for(j = jsta2-1; j < jend2; j++){
      for(i = 1; i < (m - 1); i++){
	     sx[i + j * n] = a[(i-1) + j*n] + a[(i+1)+ j*n] + 2 * a[i + j*n];
	     sz[i + j * n] = a[ i + (j-1)*n] + a[i + (j+1)*n] + 2 * a[i + j*n];
	      c[i + j * n] = (sx[i + j*n]/(dx*dx)) + (sz[i+j*n]/(dz*dz));
      }
   }

   MPI_Reduce(c, b, m * n, MPI_REAL8, MPI_SUM, 0, MPI_COMM_WORLD);

   if (myid == 0){
      show_matrix(b, n);
      //file = fopen("laplaciano_c.dat", "wb");
      //fwrite(b, sizeof(double), n * m, file);
      //fclose(file);
   }

   free(a);
   free(b);
   free(c);
   free(sx);
   free(sz);

   MPI_Finalize();

   return 0;

}/*main*/
